### ISSUE TITLE
Installation Issue

### ISSUE THREAD

**Issue by alex-stanoyevitch** [rstudio/keras3#1491]
This is my first time putting up an issue in GitHub, so I apologize in advance in case I may be doing something improperly in this submission.   

I've been reading the 2nd edition of Chollet's book for a while now and REALLY hope to start using keras + tensorflow.   
I have done following the installation instructions in Chap 3 for my PC: 
1.  First installed Python, Git, and Rtools.
2. Then I ran the following codes provided kindly to me by one of the book's authors (Tomasz K)
install.packages("keras3")
reticulate::install_python("3.11")
keras3::install_keras()

When I paste (in order) all of the RCode snippets from Chap 2 to build the first network, all seems to work fine till I run the snippet:  

model <- keras_model_sequential(list(
layer_dense(units = 512, activation = "relu"),
layer_dense(units = 10, activation = "softmax")
))

I get this error message:   
'''Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  `layers` must be passed in as a named argument.'''

I'd very much appreciate anyone's help or advice on this.   

My background:  I am a professor in math and statistics and have been teaching machine learning every year for 8 years and have been using R for 13 years.   I've never used Python nor needed to install it till I installed keras3.   I am planning a new course in deep learning, but I need to work out such bugs and issues so as to avoid my students running into them.    I need a powerful software but also a stable working environment. 
My email address is alex.stanoyevitch@gmail.com in case you need or would prefer to contact me directly on this issue.

**Comment by t-kalinowski** [comment]
Hi,

That error occurs if you are using `keras` instead of `keras3`. 

Can you please try the development version of keras3 and let me know if this works for you?

```r
pak::pak("rstudio/keras3")
print(keras3::op_convert_to_tensor("Hello World!"))
print(reticulate::py_config())
```

**Comment by alex-stanoyevitch** [comment]
Thank you Tomasz,   I ran these three commands in RStudio, it gave the following output:

pak::pak("rstudio/keras3")
'''! Using bundled GitHub PAT. Please add your own PAT using `gitcreds::gitcreds_set()`.
✔ Updated metadata database: 5.67 MB in 15 files.                          
✔ Updating metadata database ... done                                      
                                                                            
→ Will update 2 packages.
→ Will download 1 CRAN package (2.26 MB).
→ Will download 1 package with unknown size.
+ keras3     1.2.0  → 1.3.0  [bld][cmp][dl] (GitHub: e357997)
+ reticulate 1.39.0 → 1.41.0 [dl] (2.26 MB)

! keras3 and reticulate are loaded in the current R session, you probably need to restart R after the installation.

? Do you want to continue (Y/n) print(keras3::op_convert_to_tensor("Hello World!"))
Error: Aborted.
> y
 [1] -1.0 -0.9 -0.8 -0.7 -0.6 -0.5 -0.4 -0.3 -0.2 -0.1  0.0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  1.0'''

I restarted RStudio and tried running again the same code from Chap 2 of the Chollet book, but when I got to the same snippet in my original post:

model <- keras_model_sequential(list(
layer_dense(units = 512, activation = "relu"),
layer_dense(units = 10, activation = "softmax")
))

But I am still getting the same error.   

Do you have anything else that I can try?

**Comment by alex-stanoyevitch** [comment]
Added information to last post:   
I checked the list of installed packages in RStudio; keras3 is there but not keras.

**Comment by t-kalinowski** [comment]
What is the output of `py_config()`?

**Comment by alex-stanoyevitch** [comment]
> py_config()
'''Error in py_config() : could not find function "py_config"'''

**Comment by t-kalinowski** [comment]
`reticulate::py_config()`

**Comment by alex-stanoyevitch** [comment]
> reticulate::py_config()
'''python:         C:/Users/Alex/Documents/.virtualenvs/r-reticulate/Scripts/python.exe
libpython:      C:/Users/Alex/AppData/Local/Programs/Python/Python313/python313.dll
pythonhome:     C:/Users/Alex/Documents/.virtualenvs/r-reticulate
version:        3.13.1 (tags/v3.13.1:0671451, Dec  3 2024, 19:06:28) [MSC v.1942 64 bit (AMD64)]
Architecture:   64bit
numpy:          C:/Users/Alex/Documents/.virtualenvs/r-reticulate/Lib/site-packages/numpy
numpy_version:  2.2.0'''

**Comment by t-kalinowski** [comment]
Can you please run:

```r
reticulate::virtualenv_remove("r-reticulate")
```

Alternatively, you can also use the unexported `rm_all_reticulate_state` to completely revert to a clean slate:
```
reticulate:::rm_all_reticulate_state(TRUE)
```

Then try to use keras again.

**Comment by alex-stanoyevitch** [comment]
I just ran the first option, and tried to build the keras3 model again, but I am getting the same error mssg:

> reticulate::virtualenv_remove("r-reticulate")
'''Remove virtual environment 'r-reticulate'? [Y/n]: y
Virtual environment "r-reticulate" removed.'''
> model = keras_model_sequential(list(
+   layer_dense(units = 512, activation = "relu"),
+   layer_dense(units = 10, activation = "softmax")
+ ))
'''Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  could not find function "keras_model_sequential"'''

**Comment by alex-stanoyevitch** [comment]
Oh I apologize, the keras3 error message now (above) is different.  
Should I reinstall keras3 + tensorflow?

**Comment by t-kalinowski** [comment]
No need to reinstall the R package. It looks like you need to restart your R session and call `library(keras3)`

**Comment by alex-stanoyevitch** [comment]
Thank you Tomasz for all of your prompt follow-ups.  
I need to leave for work now and also since today is my birthday (I'll be celebrating after work) so will get back to this tomorrow and let you know.

**Comment by t-kalinowski** [comment]
Happy Birthday!

**Comment by alex-stanoyevitch** [comment]
Good Morning Tomasz,    I have closed all my running RStudio sessions/projects and restarted the one where I am trying to make keras3 work.   I redid the library command for keras3 (as you suggested) + for tensor flow.   The same code snippet (the model creation) gave an error mssg that was longer than before but the last part was the same as before; here is all the i/o:

> library(keras3); library(tensorflow)

'''Attaching package: ‘tensorflow’

The following objects are masked from ‘package:keras3’:

    set_random_seed, shape'''

> model = keras_model_sequential(list(
+   layer_dense(units = 512, activation = "relu"),
+   layer_dense(units = 10, activation = "softmax")
+ ))
'''2025-03-05 06:56:36.429873: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-03-05 06:56:38.208961: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  `layers` must be passed in as a named argument.'''

**Comment by alex-stanoyevitch** [comment]
I've just restarted R studio and tried the Chollet book code snipet again.  Now I am getting an earlier seen error mssg again:

> library(keras3); library(tensorflow)

'''Attaching package: ‘tensorflow’

The following objects are masked from ‘package:keras3’:

    set_random_seed, shape'''

> model = keras_model_sequential(list(
+     
+     layer_dense(units = 512, activation = "relu"),
+     layer_dense(units = 10, activation = "softmax")
+ ))
'''2025-03-05 13:44:06.200054: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-03-05 13:44:07.282291: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  `layers` must be passed in as a named argument.'''

**Comment by t-kalinowski** [comment]
The repeated messages like this:
```
Warning: cannot remove prior installation of package ‘jsonlite’
Warning: restored ‘jsonlite’
``` 

indicate that your attempt to update R packages is not working. This is sometimes an issue on Windows, where it is not possible to delete a file while another process has the file open.

`pak::pak()` has some additional handling to help with this situation. Or you can make sure to update packages only in a freshly restarted R session with no R files or projects open, before any of the packages-to-be-updated have been loaded.

By the way, you can format code chunks in github with triple backticks.

**Comment by alex-stanoyevitch** [comment]
Thankyou Tomasz,   It is good to know why such error mssgs happen and I will next (and in the future) aim to install all R packages in a fresh empty R session--I will create a new "empty" project just for such purposes.    I'll get back to you later after I try all this, and will keep my fingers crossed.   

BTW, thanks for correcting my use of the wrong tic for code formatting.

**Comment by alex-stanoyevitch** [comment]
Hi Tomasz,   I used the first method you suggested (work on an empty RStudio project) and redid those commands you gave before.   Things looked good (the packages now seemed to update correctly) so I did not save the output, closed the R session and reopened my "deep learning project" to run those commands.   But I am again getting that same error mssg:

``` library(keras3); library(tensorflow)```

Attaching package: ‘tensorflow’

The following objects are masked from ‘package:keras3’:

    set_random_seed, shape

```model = keras_model_sequential(list(
+     
+     layer_dense(units = 512, activation = "relu"),
+     layer_dense(units = 10, activation = "softmax")
+ ))```

2025-03-05 18:33:16.186836: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-03-05 18:33:17.966154: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  `layers` must be passed in as a named argument.

**Comment by alex-stanoyevitch** [comment]
```model = keras_model_sequential(list(
+     
+     layer_dense(units = 512, activation = "relu"),
+     layer_dense(units = 10, activation = "softmax")
+ ))```

2025-03-05 18:33:16.186836: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-03-05 18:33:17.966154: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Error in keras_model_sequential(list(layer_dense(units = 512, activation = "relu"),  : 
  `layers` must be passed in as a named argument.

**Comment by alex-stanoyevitch** [comment]
The whole code snippet (that I enclosed in triple back tics) did not paste completely in the above two attempts, here is the code:

> model = keras_model_sequential(list(
+     
+     layer_dense(units = 512, activation = "relu"),
+     layer_dense(units = 10, activation = "softmax")
+ ))

**Comment by t-kalinowski** [comment]
Thanks Alex, this reminded me that the arguments to `keras_model_sequential()` changed in `keras3`. You can fix by passing `layers` as a named argument `keras_model_sequential(layers = list(...))`. 

Note that that usage pattern is only used once early on for didactic purposes. The more common usage pattern is to compose with `|>`

```r
keras_model_sequential() |> 
  layer_*() |>
  layer_*() ...
```

You may want to consult https://github.com/jonbry/deep-learning-with-R-2nd-edition-code-keras3, where @jonbry has kindly updated all code examples from the book for `keras3`
